src/modules/accounting/accounting.py

```diff
@@ -327,3 +312,103 @@ def get_extra_data(self, blockstamp: ReferenceBlockStamp) -> ExtraData:
+        rebase_part = self._calculate_rebase_report(blockstamp)
+        modules_part = self._get_newly_exited_validators_by_modules(blockstamp)
+        wq_part = self._calculate_wq_report(blockstamp)
+
+        extra_data_part_v2 = self._calculate_extra_data_report_v2(blockstamp)
+        return self._combine_report_parts(2, blockstamp, rebase_part, modules_part, wq_part, extra_data_part_v2)
+
+    # fetches validators_count, cl_balance, withdrawal_balance, el_vault_balance, shares_to_burn
+    def _calculate_rebase_report(self, blockstamp: ReferenceBlockStamp) -> tuple[int, Gwei, Wei, Wei, int]:
+        validators_count, cl_balance = self._get_consensus_lido_state(blockstamp)
+        withdrawal_vault_balance = self.w3.lido_contracts.get_withdrawal_balance(blockstamp)
+        el_rewards_vault_balance = self.w3.lido_contracts.get_el_vault_balance(blockstamp)
+        shares_requested_to_burn = self.get_shares_to_burn(blockstamp)
+        return validators_count, cl_balance, withdrawal_vault_balance, el_rewards_vault_balance, shares_requested_to_burn
+
+    # calculates is_bunker, finalization_share_rate, finalization_batches
+    def _calculate_wq_report(self, blockstamp: ReferenceBlockStamp) -> tuple[bool, int, list[int]]:
+        is_bunker = self._is_bunker(blockstamp)
+        finalization_share_rate, finalization_batches = self._get_finalization_data(blockstamp)
+        return is_bunker, finalization_share_rate, finalization_batches
+
+    def _calculate_extra_data_report_v1(self, blockstamp: ReferenceBlockStamp) -> ExtraData:
+        stuck_validators, exited_validators, orl = self._get_generic_extra_data(blockstamp)
+        return ExtraDataService.collect(
+            stuck_validators,
+            exited_validators,
+            orl.max_items_per_extra_data_transaction,
+            orl.max_node_operators_per_extra_data_item,
+        )
+
+    def _calculate_extra_data_report_v2(self, blockstamp: ReferenceBlockStamp) -> ExtraData:
+        stuck_validators, exited_validators, orl = self._get_generic_extra_data(blockstamp)
+        return ExtraDataServiceV2.collect(
+            stuck_validators,
+            exited_validators,
+            orl.max_items_per_extra_data_transaction,
+            orl.max_node_operators_per_extra_data_item,
+        )
+
+    @staticmethod
+    def _update_metrics(report_data: ReportData):
+        ACCOUNTING_IS_BUNKER.set(report_data.is_bunker)
+        ACCOUNTING_CL_BALANCE_GWEI.set(report_data.cl_balance_gwei)
+        ACCOUNTING_EL_REWARDS_VAULT_BALANCE_WEI.set(report_data.el_rewards_vault_balance)
+        ACCOUNTING_WITHDRAWAL_VAULT_BALANCE_WEI.set(report_data.withdrawal_vault_balance)
+
+    @staticmethod
+    def _combine_report_parts(
+        consensus_version: int,
+        blockstamp: ReferenceBlockStamp,
+        report_rebase_part: tuple[int, Gwei, Wei, Wei, int],
+        report_modules_part: tuple[list[StakingModuleId], list[int]],
+        report_wq_part: tuple[bool, int, list[int]],
```

Comment:
This looks like a bottleneck in our typing - perhaps it would be nice to create a separate type for each int or wei and then it would be impossible to mix them up.

Currently, it's not understandable tuple[int, Gwei, Wei, Wei, int] - what params it is. In some bad cases we can tuple[2] instead of tuple[3] and typings won't help us.

Reply:
Done